Add function pushdown query plan optimization

talmawash opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Optimize queries that use volume reducing functions like length and substr, closes https://github.com/ClickHouse/ClickHouse/issues/82378

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


### Optimization Breakdown
The optimization conflicts with other optimizations of the 1st pass that do the opposite of function pushdown, so we run in the 2nd pass. 

#### Helper functions
1. Build a map for child -> parent relations for the plan
2. Find source/reading node

The results from the helper functions are passed to the `processNodeDFS` function, which is the bulk of the optimization. Here's how it works generally in a step-by-step breakdown:

#### Step 1 - Analysis

- Analyze the inputs and outputs of each node
  - For columns/functions seen for the first time, we just create a new record
  - For existing columns, we update the input and output names for the particular node, and if the column is used in a way that affects push down (e.g., unsupported function, sorting, joins), we set the max pushdown point to the current node

#### Step 2 - Request Creation
- Create pushdown requests for each column, grouping functions by column
  - Only suitable functions with a single input column (constants are fine) are added

#### Step 3 - Pushdown
- If the current node is the max pushdown node, push down to it
  - Find the first expression/filter (inclusive) step node at or above to call the function
  - Remove the input column from outputs
  - Update input headers of the parent nodes

### PR Todos:
- [ ] Look into refactoring or applying a simpler approach (review feedback affects this)
- [x] Explain the approach
- [ ] Validate any assumptions
- [x] Ensure order is consistent and queries don't break when using the optimization
- [x] Support more functions
- [x] Add functional and performance tests

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new query plan optimization that rewrites `Expression`/`Filter` DAGs and can insert new `ExpressionStep`s (including around `JOIN`s), which may affect correctness or performance in edge cases. Guarded by a new setting and includes functional/perf coverage, but touches core planner/optimizer behavior.
> 
> **Overview**
> Adds a new query-plan optimization, gated by `query_plan_push_down_volume_reducing_functions`, that **pushes “volume-reducing” functions (e.g. `length`, `empty`/`notEmpty`, `substring`-related search) closer to sources** to reduce data movement before operations like sorting.
> 
> Introduces `IFunction::isSuitableForPushDownBeforeFilter()` and marks several string/array functions as pushdown-eligible; updates `liftUpFunctions` to avoid lifting these above `SortingStep` to prevent optimization ping-pong.
> 
> Implements `tryPushDownVolumeReducingFunctions` (including special handling for `JoinStepLogical` via a new `addPassThroughColumn`) and adds stateless + performance tests validating identical results with the optimization on/off.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ff96f25770423c2fe621a3ac4d27eaa080e2cb56. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

talmawash committed:
add func pushdown settings

talmawash committed:
add func pushdown suitability property

talmawash committed:
add func pushdown optimization

talmawash committed:
support empty pushdown

talmawash committed:
refactor function pushdown

talmawash committed:
conflict resolve fix

talmawash committed:
style fixes

talmawash committed:
fix quotes

talmawash committed:
temp enable pushdown optimization for tests

talmawash committed:
handle function not projected case

talmawash committed:
add basic test

talmawash committed:
fix bug

talmawash commented:
Right now, what I'm working on is replicating the issue since some of these tests are actually passing locally

I'm speculating here, but the ones I couldn't run easily due to the test setup might be the culprit for the failure through some side effects.

talmawash committed:
remove original stale call

talmawash committed:
add tests

talmawash committed:
doc fix

talmawash committed:
minor improvements

talmawash committed:
support pushdown for more functions

talmawash committed:
use dfs instead of recursion

talmawash committed:
add accidentally removed setting

talmawash commented on src/Processors/QueryPlan/Optimizations/pushDownVolumeReducingFunctions.cpp:
```diff
@@ -0,0 +1,781 @@
+            column_to_requests[request->input_column_candidate][parent_node] = request;
+    }
+
+    // only return requests shared by all parents
```
Maybe push down all the way up to the parents that have it?

novikd commented on src/Core/SettingsChangesHistory.cpp:
```diff
@@ -74,6 +74,7 @@ const VersionToSettingsChangesMap & getSettingsChangesHistory()
             {"os_threads_nice_value_query", 0, 0, "New setting."},
             {"os_threads_nice_value_materialized_view", 0, 0, "New setting."},
             {"os_thread_priority", 0, 0, "Obsolete setting."},
+            {"query_plan_push_down_volume_reducing_functions", false, false, "A new setting"},
```
```suggestion
            {"query_plan_push_down_volume_reducing_functions", false, true, "A new setting"},
```

novikd commented on src/Processors/QueryPlan/Optimizations/pushDownVolumeReducingFunctions.cpp:
```diff
@@ -0,0 +1,781 @@
+};
+
+/// Represents a request to push down a volume-reducing function
+struct PushdownRequest
```
This is also not needed.

novikd reviewed (commented):
The implementation is currently overcomplicated. It would be much simpler and less code if only 2 query plan steps were analyzed at a time.

talmawash commented on src/Processors/QueryPlan/Optimizations/pushDownVolumeReducingFunctions.cpp:
```diff
@@ -0,0 +1,781 @@
+struct ColumnCandidate
+{
+    /// This represents how far back functions using this node can be pushed down
+    QueryPlan::Node * max_pushdown_point;
```
I think what we can do is handle cases like the join case attached above with a separate optimization, wdyt?

novikd committed:
Merge remote-tracking branch 'origin/master' into add_function_pushdown

novikd committed:
WIP on rewriting

novikd committed:
Move optimization to the first pass

novikd committed:
Clean up code

novikd committed:
Merge remote-tracking branch 'origin/master' into add_function_pushdown

novikd committed:
Try to fix crash

novikd committed:
WIP

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into add_function_pushdown

# Conflicts:
#	src/Processors/QueryPlan/Optimizations/Optimizations.h
#	src/Processors/QueryPlan/Optimizations/liftUpFunctions.cpp

alexey-milovidov committed:
Fix style: remove space before closing paren in for-loop

The CI style check failed on two for-loops with `; )` — the check rejects
whitespace before `)`.

See https://github.com/ClickHouse/ClickHouse/pull/86139

alexey-milovidov committed:
Fix settings history for `query_plan_push_down_volume_reducing_functions`

Two issues called out by review feedback on PR #86139:

- The actual default declared in `Settings.cpp` is `true`, but the
  changes-history entry recorded `{false, false}`. Fixed to
  `{false, true}` per @novikd's review suggestion.
- A duplicate `os_thread_priority` entry was accidentally added during
  a rebase; the canonical entry "Alias for os_threads_nice_value_query."
  is preserved.

See https://github.com/ClickHouse/ClickHouse/pull/86139

alexey-milovidov committed:
Fix inverted `ANTI` join pushdown sides

For `LEFT ANTI` / `RIGHT ANTI`, output rows are taken from the
preserved side only (no columns from the eliminated side flow to the
output), so pushdown should target the preserved side — the same side
already used for non-`ANTI` `LEFT`/`RIGHT`. The previous logic flipped
`push_to_left`/`push_to_right` for `ANTI`, which silently disabled the
optimization on the side where it could actually fire.

Reported by cursor bugbot on PR #86139.

alexey-milovidov committed:
Rename `isSuitableForPushDownBeforeFilter` to `isVolumeReducing`

The previous name suggested the property was tied to filter pushdown,
but the optimization is about volume-reducing functions in general
(it does not depend on the presence of a filter). Renamed to reflect
the actual semantics, addressing review feedback from @novikd on
PR #86139.

The template parameter `is_suitable_for_push_down_before_filter` in
`FunctionStringOrArrayToT` is renamed similarly to `is_volume_reducing`.

alexey-milovidov committed:
Restore submodule pointers to merge state

The previous commit (`isVolumeReducing` rename) accidentally bumped
submodule pointers because the worktree had submodules at stale SHAs
left over from before the merge. Restore them to the values introduced
by the master merge.

See https://github.com/ClickHouse/ClickHouse/pull/86139

novikd commented:
There's another PR with a simpler approach https://github.com/ClickHouse/ClickHouse/pull/106199.
